src/lib/location-provider.js

```diff
@@ -0,0 +1,49 @@
+import * as React from "react"
+import { globalHistory, navigate } from "./history"
+import { isRedirect } from "./redirect"
+import { LocationContext } from "./hooks-create-context"
+import { withErrorBoundary, useErrorBoundary } from "./error-boundary"
+
+export const LocationProvider = withErrorBoundary(
+  ({ history = globalHistory, children }) => {
+    const { location } = history
+
+    const [context, setContext] = React.useState({ location })
+    const [refs, setRefs] = React.useState({ unlisten: null })
+    const [error, _] = useErrorBoundary()
+
+    React.useEffect(() => {
+      history._onTransitionComplete()
+    }, [context.location])
+
+    React.useEffect(() => {
+      history._onTransitionComplete()
+      const unlisten = history.listen(() => {
+        Promise.resolve().then(() => {
+          requestAnimationFrame(() => {
+            setContext({ location })
+          })
+        })
+      })
+      setRefs({ unlisten })
+
+      return () => {
+        refs.unlisten()
+      }
+    }, [])
```

Comment:
No need to setState for unlisten as it's only used in scope.  I've also added an isCancelled to make sure React don't throw on unmount because setContext is async.

```suggestion
    React.useEffect(() => {
      let isCancelled = false;
      const unlisten = history.listen(() => {
        Promise.resolve().then(() => {
          requestAnimationFrame(() => {
            if (!isCancelled) {
              setContext({ location })
            }
          })
        })
      })
      return () => {
        isCancelled = true;
        unlisten()
      }
    }, [])
```